### ISSUE TITLE
descriptor generating incorrect output for arrow functions with implicit typing returning this

### ISSUE THREAD

**Issue by grofit** [Microsoft/TypeScript#8188]
I have looked and cannot find this raised already and asked in gitter and no one seemed to know.

So if I had the code like below:

```ts
export class SomeBuilder
{
   public something = () => { return this; }
}
```

which seems to output the d.ts file like so:

```ts
export declare class SomeBuilder{
    something: () => this;
}
```

Which is incorrect as `this` is not a type so it blows up, if I explicitly change it to:

```ts
public something = (): SomeBuilder => { return this; }
```

It outputs the type as expected, If this is already known or expected behaviour I apologise.

**Comment by mhegazy** [comment]
> Which is incorrect as this is not a type so it blows up

Why? `this` is a valid type name, and it refers to the type of `this` in an expression position. what version of TS are you using to consume the .d.ts output?

**Comment by mhegazy** [comment]
see http://www.typescriptlang.org/docs/handbook/advanced-types.html#polymorphic-this-types for more documentation about `this` types.

**Comment by grofit** [comment]
1.8.9 it is being consumed via `gulp-typescript` which blows up with TSC errors when trying to include the d.ts file with the `this` type, however when I set it to the actual type it works fine. I am not on the office PC but I can get you the actual output tomorrow if that helps.

**Comment by mhegazy** [comment]
If you have an explicit type annotation, it will **always** be used in the output. if you do not have one, the compiler tries to infer one for you. in this case, your function is returning `this`, whose type is not `SomeBuilder`, but rather "`SomeBuilder` or one of its descendants", i.e. `this`.

I am assuming that your gulp-typescript is using an older version of typescript that does not understand this types.

**Comment by grofit** [comment]
Hello, you are 100% correct, turns out it was actually type-doc which was throwing the errors, it just *looked* on the output log like it was from the normal typescript compilaiton process. I am using explicit types for the moment but its good to know that `this` is acceptable in the new world.
